<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
    <title>RNAseq Using R </title>
    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="css/swc.css" />
    <meta charset="UTF-8" />
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body class="lesson">
    <div class="container card">
      <div class="banner" align="right" style="margin-top: 5px">
        <a href="http://monash.edu/bioinformatics" title="Monash Bioinformatics Platform">
            <div style="margin-right: 5px;" class="label swc-blue-bg">Monash Bioinformatics Platform</div>
        </a>
      </div>
      <article>
      <div class="row">
        <div class="col-md-10 col-md-offset-1">
                    <h1 class="title"></h1>
          <h1 id="rnaseq-de-analysis-with-r---solutions">RNAseq DE analysis with R - Solutions</h1>
<p>This document provides solutions to the exercises given in the ‘RNAseq DE analysis with R’ course.</p>
<h2 id="clustering">Clustering</h2>
<pre><code>## Warning in readChar(con, 5L, useBytes = TRUE): cannot open compressed file
## &#39;/home/roxane/RNAseq_DE_analysis_with_R.RData&#39;, probable reason &#39;No such
## file or directory&#39;</code></pre>
<pre><code>## Error in readChar(con, 5L, useBytes = TRUE): cannot open the connection</code></pre>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2><span class="glyphicon glyphicon-pencil"></span>Exercise: Heatmap</h2>
</div>
<div class="panel-body">
<p>Produce a heatmap for the 50 most highly expressed genes and annotate the samples with with their age.</p>
<ul>
<li>Subset the read counts object for the 50 most highly expressed genes</li>
<li>Annotate the samples in the subset with their age (check order with design!)</li>
<li>Plot a heatmap with this subset of data, scaling genes and ordering both genes and samples</li>
</ul>
</div>
</section>
<p>Solution:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Subset the read counts object for the 30 most highly expressed genes</span>
select =<span class="st"> </span><span class="kw">order</span>(<span class="kw">rowMeans</span>(counts$counts), <span class="dt">decreasing=</span><span class="ot">TRUE</span>)[<span class="dv">1</span>:<span class="dv">50</span>]</code></pre>
<pre><code>## Error in is.data.frame(x): object &#39;counts&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">highexprgenes_counts &lt;-<span class="st"> </span>counts$counts[select,]</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;counts&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Annotate the samples in the subset with their age (check order with design!)</span>
<span class="kw">colnames</span>(highexprgenes_counts)&lt;-<span class="st"> </span>experiment_design.ord$age</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;experiment_design.ord&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(highexprgenes_counts)</code></pre>
<pre><code>## Error in head(highexprgenes_counts): object &#39;highexprgenes_counts&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot a heatmap with this subset of data, scaling genes and ordering both genes and samples</span>
<span class="kw">heatmap</span>(highexprgenes_counts, <span class="dt">col=</span><span class="kw">topo.colors</span>(<span class="dv">50</span>), <span class="dt">margin=</span><span class="kw">c</span>(<span class="dv">10</span>,<span class="dv">6</span>))</code></pre>
<pre><code>## Error in heatmap(highexprgenes_counts, col = topo.colors(50), margin = c(10, : object &#39;highexprgenes_counts&#39; not found</code></pre>
<h2 id="principal-component-analysis">Principal Component Analysis</h2>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2><span class="glyphicon glyphicon-pencil"></span>Exercise: PCA</h2>
</div>
<div class="panel-body">
<p>Produce a PCA plot from the read counts of the 500 most highly expressed genes and change the labels until you can identify the reason for the split between samples from the same tissue.</p>
<ul>
<li>Get the read counts for the 500 most highly expressed genes</li>
<li>Transpose this matrix of read counts</li>
<li>Check the number of dimensions explaining the variability in the dataset</li>
<li>Run the PCA with an appropriate number of components</li>
<li>Annotate the samples with their age &amp; re-run the PCA &amp; plot the main components</li>
<li>Annotate the samples with other clinical data &amp; re-run the PCA &amp; plot the main components until you can separate the samples within each tissue group</li>
</ul>
</div>
</section>
<p>Solution:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select data for the 1000 most highly expressed genes</span>
select =<span class="st"> </span><span class="kw">order</span>(<span class="kw">rowMeans</span>(counts$counts), <span class="dt">decreasing=</span><span class="ot">TRUE</span>)[<span class="dv">1</span>:<span class="dv">500</span>]</code></pre>
<pre><code>## Error in is.data.frame(x): object &#39;counts&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">highexprgenes_counts &lt;-<span class="st"> </span>counts$counts[select,]</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;counts&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># transpose the data to have variables (genes) as columns</span>
data_for_PCA &lt;-<span class="st"> </span><span class="kw">t</span>(highexprgenes_counts)</code></pre>
<pre><code>## Error in t(highexprgenes_counts): object &#39;highexprgenes_counts&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Run the PCA with an appropriate number of components</span>
mds &lt;-<span class="st"> </span><span class="kw">cmdscale</span>(<span class="kw">dist</span>(data_for_PCA))</code></pre>
<pre><code>## Error in as.matrix(x): object &#39;data_for_PCA&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot the PCA</span>
<span class="kw">plot</span>(mds[,<span class="dv">1</span>], -mds[,<span class="dv">2</span>], <span class="dt">type=</span><span class="st">&quot;n&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Dimension 1&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Dimension 2&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>)</code></pre>
<pre><code>## Error in plot(mds[, 1], -mds[, 2], type = &quot;n&quot;, xlab = &quot;Dimension 1&quot;, ylab = &quot;Dimension 2&quot;, : object &#39;mds&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">text</span>(mds[,<span class="dv">1</span>], -mds[,<span class="dv">2</span>], <span class="kw">rownames</span>(mds), <span class="dt">cex=</span><span class="fl">0.8</span>) </code></pre>
<pre><code>## Error in text(mds[, 1], -mds[, 2], rownames(mds), cex = 0.8): object &#39;mds&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Annotate the samples with their age &amp; re-run the PCA &amp; plot the main components</span>
<span class="kw">rownames</span>(mds) &lt;-<span class="st"> </span>experiment_design.ord$age</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;experiment_design.ord&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(mds[,<span class="dv">1</span>], -mds[,<span class="dv">2</span>], <span class="dt">type=</span><span class="st">&quot;n&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Dimension 1&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Dimension 2&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>)</code></pre>
<pre><code>## Error in plot(mds[, 1], -mds[, 2], type = &quot;n&quot;, xlab = &quot;Dimension 1&quot;, ylab = &quot;Dimension 2&quot;, : object &#39;mds&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">text</span>(mds[,<span class="dv">1</span>], -mds[,<span class="dv">2</span>], <span class="kw">rownames</span>(mds), <span class="dt">cex=</span><span class="fl">0.8</span>) </code></pre>
<pre><code>## Error in text(mds[, 1], -mds[, 2], rownames(mds), cex = 0.8): object &#39;mds&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Annotate the samples with other clinical data &amp; re-run the PCA &amp; plot the main components until you can separate the samples within each tissue group</span>
<span class="kw">rownames</span>(mds)&lt;-<span class="st"> </span>experiment_design.ord$technical_replicate_group</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;experiment_design.ord&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(mds[,<span class="dv">1</span>], -mds[,<span class="dv">2</span>], <span class="dt">type=</span><span class="st">&quot;n&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;Dimension 1&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Dimension 2&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>)</code></pre>
<pre><code>## Error in plot(mds[, 1], -mds[, 2], type = &quot;n&quot;, xlab = &quot;Dimension 1&quot;, ylab = &quot;Dimension 2&quot;, : object &#39;mds&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">text</span>(mds[,<span class="dv">1</span>], -mds[,<span class="dv">2</span>], <span class="kw">rownames</span>(mds), <span class="dt">cex=</span><span class="fl">0.8</span>) </code></pre>
<pre><code>## Error in text(mds[, 1], -mds[, 2], rownames(mds), cex = 0.8): object &#39;mds&#39; not found</code></pre>
<h2 id="differential-expression">Differential Expression</h2>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2><span class="glyphicon glyphicon-pencil"></span>Exercise: Limma</h2>
</div>
<div class="panel-body">
<p>Get the number of DE genes between technical group 1 and technical group 2 (all Brain samples) with adj pvalue&lt;0.01.</p>
<ul>
<li>Create a new design matrix for limma with the technical replicate groups</li>
<li>Re-normalise the read counts with ‘voom’ function with new design matrix</li>
<li>Fit a linear model on these normalised data</li>
<li>Make the contrast matrix corresponding to the new set of parameters</li>
<li>Fit the contrast matrix to the linear model</li>
<li>Compute moderated t-statistics of differential expression</li>
<li>Get the output table for the 10 most significant DE genes for this comparison</li>
</ul>
</div>
</section>
<p>Solution:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(limma)</code></pre>
<pre><code>## Loading required package: methods</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create a new design matrix for limma with the technical replicate groups</span>
techgroup&lt;-<span class="kw">factor</span>(experiment_design.ord$technical_replicate_group)</code></pre>
<pre><code>## Error in factor(experiment_design.ord$technical_replicate_group): object &#39;experiment_design.ord&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(~<span class="dv">0</span>+techgroup)</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;techgroup&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(design)&lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;techgroup&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="kw">colnames</span>(design))</code></pre>
<pre><code>## Error in is.data.frame(x): object &#39;design&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">design</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;design&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Re-normalise the read counts with &#39;voom&#39; function with new design matrix</span>
y &lt;-<span class="st"> </span><span class="kw">voom</span>(mycounts,design,<span class="dt">lib.size=</span><span class="kw">colSums</span>(mycounts)*nf)</code></pre>
<pre><code>## Error in is(counts, &quot;DGEList&quot;): object &#39;mycounts&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">counts.voom &lt;-<span class="st"> </span>y$E</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;y&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fit a linear model on these normalised data</span>
fit &lt;-<span class="st"> </span><span class="kw">lmFit</span>(y,design)</code></pre>
<pre><code>## Error in is(object, &quot;list&quot;): object &#39;y&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Make the contrast matrix corresponding to the new set of parameters</span>
cont.matrix &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>(group_2-group_1,<span class="dt">levels=</span>design)</code></pre>
<pre><code>## Error in is.factor(levels): object &#39;design&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">cont.matrix </code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;cont.matrix&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fit the contrast matrix to the linear model</span>
fit &lt;-<span class="st"> </span><span class="kw">contrasts.fit</span>(fit, cont.matrix)</code></pre>
<pre><code>## Error in NCOL(fit$coefficients): object &#39;fit&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Compute moderated t-statistics of differential expression </span>
fit &lt;-<span class="st"> </span><span class="kw">eBayes</span>(fit)</code></pre>
<pre><code>## Error in ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim, : object &#39;fit&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">options</span>(<span class="dt">digits=</span><span class="dv">3</span>)

<span class="co"># Get the output table for the 10 most significant DE genes for this comparison</span>
<span class="kw">dim</span>(<span class="kw">topTable</span>(fit,<span class="dt">coef=</span><span class="st">&quot;group_2 - group_1&quot;</span>,<span class="dt">p.val=</span><span class="fl">0.01</span>,<span class="dt">n=</span><span class="ot">Inf</span>))</code></pre>
<pre><code>## Error in is(fit, &quot;MArrayLM&quot;): object &#39;fit&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">topTable</span>(fit,<span class="dt">coef=</span><span class="st">&quot;group_2 - group_1&quot;</span>,<span class="dt">p.val=</span><span class="fl">0.01</span>)</code></pre>
<pre><code>## Error in is(fit, &quot;MArrayLM&quot;): object &#39;fit&#39; not found</code></pre>
<h2 id="gene-ontology">Gene Ontology</h2>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2><span class="glyphicon glyphicon-pencil"></span>Exercise: GOstats</h2>
</div>
<div class="panel-body">
<p>Identify the GO terms in the Molecular Function domain that are over-represented (pvalue&lt;0.01) in your list of DE genes.</p>
<ul>
<li>Get your list of DE genes (Entrez Gene IDs)</li>
<li>Set the new parameters for the hypergeometric test</li>
<li>Run the test and adjust the pvalues in the output object</li>
<li>Identify the significant GO terms at pvalue 0.01</li>
</ul>
</div>
</section>
<p>Solution:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GOstats)</code></pre>
<pre><code>## Error in library(GOstats): there is no package called &#39;GOstats&#39;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get your list of DE genes (Entrez Gene IDs)</span>
entrezgeneids &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">rownames</span>(limma.res.pval.FC))</code></pre>
<pre><code>## Error in rownames(limma.res.pval.FC): object &#39;limma.res.pval.FC&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">universeids &lt;-<span class="st"> </span><span class="kw">rownames</span>(mycounts)</code></pre>
<pre><code>## Error in rownames(mycounts): object &#39;mycounts&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Set the new parameters for the hypergeometric test</span>
params &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;GOHyperGParams&quot;</span>,<span class="dt">annotation=</span><span class="st">&quot;org.Hs.eg&quot;</span>,<span class="dt">geneIds=</span>entrezgeneids,<span class="dt">universeGeneIds=</span>universeids,<span class="dt">ontology=</span><span class="st">&quot;MF&quot;</span>,<span class="dt">pvalueCutoff=</span><span class="fl">0.01</span>,<span class="dt">testDirection=</span><span class="st">&quot;over&quot;</span>)</code></pre>
<pre><code>## Error in getClass(Class, where = topenv(parent.frame())): &quot;GOHyperGParams&quot; is not a defined class</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Run the test and adjust the pvalues in the output object</span>
hg &lt;-<span class="st"> </span><span class="kw">hyperGTest</span>(params)</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): could not find function &quot;hyperGTest&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">hg.pv &lt;-<span class="st"> </span><span class="kw">pvalues</span>(hg)</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): could not find function &quot;pvalues&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">hg.pv.fdr &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(hg.pv,<span class="st">&#39;fdr&#39;</span>)</code></pre>
<pre><code>## Error in p.adjust(hg.pv, &quot;fdr&quot;): object &#39;hg.pv&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Identify the significant GO terms at pvalue 0.01</span>
sigGO.ID &lt;-<span class="st"> </span><span class="kw">names</span>(hg.pv.fdr[hg.pv.fdr &lt;<span class="st"> </span>hgCutoff])</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;hg.pv.fdr&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">df &lt;-<span class="st"> </span><span class="kw">summary</span>(hg)</code></pre>
<pre><code>## Error in summary(hg): object &#39;hg&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">GOterms.sig &lt;-<span class="st"> </span>df[df[,<span class="dv">1</span>] %in%<span class="st"> </span>sigGO.ID,<span class="st">&quot;Term&quot;</span>]</code></pre>
<pre><code>## Error in df[, 1]: object of type &#39;closure&#39; is not subsettable</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(GOterms.sig )</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;GOterms.sig&#39; not found</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(GOterms.sig)</code></pre>
<pre><code>## Error in head(GOterms.sig): object &#39;GOterms.sig&#39; not found</code></pre>
<h2 id="r-environment">R environment</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre>
<pre><code>## R version 3.2.0 (2015-04-16)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.1 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] methods   stats     graphics  grDevices utils     datasets  base     
## 
## other attached packages:
## [1] limma_3.16.7
## 
## loaded via a namespace (and not attached):
## [1] formatR_1.4   tools_3.2.0   knitr_1.13    stringr_0.6.2 evaluate_0.9</code></pre>
        </div>
      </div>
      </article>
      <div class="footer">
        <a class="label swc-blue-bg" href="https://github.com/MonashBioinformaticsPlatform/RNAseq-DE-analysis-with-R">Source</a>
      </div>
    </div>
    <!-- Javascript placed at the end of the document so the pages load faster -->
    <script src="http://code.jquery.com/jquery-1.9.1.js"></script>
    <script src="css/bootstrap/bootstrap-js/bootstrap.js"></script>
  </body>
</html>
