RNAseq Analysis Using R Course
=====================

Changes to this repository show on the web at [http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R](http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R)

You can teach your lessons from there.


## Contributing

Thanks to [Software Carpentry](http://software-carpentry.org/) for allowing us to take their lesson template and modify it for this cousre.

**These lessons show 'live' computed output. This means that to edit the lessons, you need to do so on a computer that has all the lesson's required R packages installed.**

When editing topic pages for R lessons, you should change the source R Markdown
file (*.Rmd).

To build the lessons themselves, render the html pages by running `make preview` from the base of the repository. Generating the html file(s) is required for viewing the online version of the lessons. Building the rendered page with the Makefile requires installing some dependencies first (see Required Packages, Environment below). This includes the R package knitr, which is used to render the html lesson pages. 

`Rscript -e 'install.packages("knitr", repos="http://cran.us.r-project.org")'`

Once you've made your edits and rendered the corresponding html files,
you need to add, commit, and push the lesson files 
and any supporting files (e.g. data files). Changes generated by the `make preview` command should not be committed or included in a pull request to us. These changes will be taken care off by the lesson maintainer when the pull-request is merged.


## Required data

Most of the required data is included in this repository, so nothing needs to be done.

**The exception is the hg19 genome reference data.**

This data can be downloaded after you download this repository by running the included script.

```
cd data/ref_data
./download_hg19.sh
```

Note: It's 8.4 gigabytes in size so it may take a while.


## Required Packages, Environment

Our lesson was built using the following R environment. This and the lesson itself should help with required packages and versions.

```
## R version 3.2.2 (2015-08-14)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.2 LTS
## 
## locale:
##  [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
##  [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
##  [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] xtable_1.8-0         org.Hs.eg.db_3.0.0   GOstats_2.32.0      
##  [4] graph_1.44.1         Category_2.32.0      GO.db_3.0.0         
##  [7] RSQLite_1.0.0        DBI_0.3.1            AnnotationDbi_1.28.2
## [10] GenomeInfoDb_1.2.5   IRanges_2.0.1        S4Vectors_0.4.0     
## [13] Matrix_1.2-3         Biobase_2.26.0       BiocGenerics_0.12.1 
## [16] biomaRt_2.22.0       edgeR_3.8.6          limma_3.22.7        
## [19] Rsubread_1.16.1     
## 
## loaded via a namespace (and not attached):
##  [1] knitr_1.11            magrittr_1.5          splines_3.2.2        
##  [4] lattice_0.20-33       stringr_1.0.0         tools_3.2.2          
##  [7] grid_3.2.2            AnnotationForge_1.8.2 genefilter_1.48.1    
## [10] survival_2.38-3       RBGL_1.42.0           GSEABase_1.28.0      
## [13] formatR_1.2.1         bitops_1.0-6          RCurl_1.95-4.7       
## [16] evaluate_0.8          stringi_1.0-1         XML_3.98-1.3         
## [19] annotate_1.44.0
```